### PR TITLE
cava 0.7.4 (new formula)

### PR DESCRIPTION
**Pull request by chincheta0815**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/cava.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  depends_on "libtool"
+  depends_on "portaudio"
+
+  uses_from_macos "ncurses"
+
+  def install
+    # force autogen.sh to look for and use our glibtoolize
+    inreplace "autogen.sh", "libtoolize", "glibtoolize"
+
+    system "./autogen.sh"
```
Comment (SMillerDev) [inline Formula/cava.rb:21]: Why do we need to re-run autogen? Doesn't upstream ship a configure?

File: Formula/cava.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    ENV.append "CFLAGS", "-g -O0"
+    system "./configure", *std_configure_args, "--disable-silent-rules"
```
Comment (SMillerDev) [inline Formula/cava.rb:23]: This seems to ignore the `--prefix` setting

File: Formula/cava.rb (hunk 1 above)
Comment (chincheta0815) [reply Formula/cava.rb:21]: Upstream does not ship a configure. They even tell to run autogen.sh a step in installation/building.

File: Formula/cava.rb (hunk 2 above)
Comment (chincheta0815) [reply Formula/cava.rb:23]: Can you specify why it seems to do so?
Line 21 is needed as stated above.
Line 22 is a stupid bug I have to remove. Sorry for that.
Line 23 is more or less the example from the Formula template.

I found an issue on the arm64 machines:
-L/usr/local/lib is not found. This is probably due to some hardcoded LDFLAGS in upstreams build routines that I have to fix. Can you confirm that?

Comment (carlocab) [reply Formula/cava.rb:23]: It seem to install into the right place though. It just really wants to still look in `/usr/local/lib` for some reason.

File: Formula/cava.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  depends_on "portaudio"
+
+  uses_from_macos "ncurses"
+
+  def install
+    # force autogen.sh to look for and use our glibtoolize
+    inreplace "autogen.sh", "libtoolize", "glibtoolize"
+
+    system "./autogen.sh"
+    ENV.append "CFLAGS", "-g -O0"
```
Comment (carlocab) [inline Formula/cava.rb:22]: Why is this needed? This won't make it past the `configure` script. The shims strip these flags after that.

File: Formula/cava.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    # `test do` will create, run in and delete a temporary directory.
+    #
+    # This test will fail and we won't accept that! For Homebrew/homebrew-core
+    # this will need to be a test that verifies the functionality of the
+    # software. Run the test with `brew test cava`. Options passed
+    # to `brew install` such as `--HEAD` also need to be provided to `brew test`.
+    #
+    # The installed folder is not in the path, so use the entire path to any
+    # executables being tested: `system "#{bin}/program", "do", "something"`.
+    assert_match "cava", shell_output("#{bin}/cava -v").strip
```
Comment (carlocab) [inline Formula/cava.rb:38]:
```suggestion
    assert_match "cava", shell_output("#{bin}/cava -v").strip
```
What happens if you try something more substantial? Can we attempt it anyway, try to get it to dump some logs and check that?

Comment (chincheta0815) [reply Formula/cava.rb]: cava has just two command line options:
  -v for printing the version
  -p for settings the path to the config file

It has an config option to send output, but this output is a raw printed audio output stream to stdout which.

Everything except "-v" is not reproducable...

File: Formula/cava.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  homepage "https://github.com/karlstav/cava"
+  url "https://github.com/karlstav/cava/archive/0.7.4.tar.gz"
+  sha256 "fefd3cc04d41b03ca416630cafadbfda6c75e2ca0869da1f03963dcb13e1ecb7"
+  license "MIT"
+  head "https://github.com/karlstav/cava.git", branch: "master"
+
+  depends_on "automake"
+  depends_on "fftw"
+  depends_on "iniparser"
+  depends_on "libtool"
```
Comment (carlocab) [inline Formula/cava.rb:12]: Are `automake` and `libtool` needed at runtime?

File: Formula/cava.rb (hunk 4 above)
Comment (carlocab) [reply Formula/cava.rb]: > -p for settings the path to the config file

We can check this too.

> It has an config option to send output, but this output is a raw printed audio output stream to stdout which.

Let's do this. We can at least check that doesn't exit with an error, which is what using `system` already does.

File: Formula/cava.rb (hunk 5 above)
Comment (chincheta0815) [reply Formula/cava.rb:12]: No. At buildtime.

Probably I misunderstood depends_on there.
Can I savely ignore depends_on for libtool and automake that when building?

Comment (carlocab) [reply Formula/cava.rb]: Sorry, I don't understand your question. Could you rephrase it?

File: Formula/cava.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  depends_on "iniparser"
+  depends_on "portaudio"
+
+  uses_from_macos "ncurses"
+
+  def install
+    # force autogen.sh to look for and use our glibtoolize
+    inreplace "autogen.sh", "libtoolize", "glibtoolize"
+    # correct Makefile.am with hardcoded libpath
+    inreplace "Makefile.am", "/usr/local", HOMEBREW_PREFIX
```
Comment (carlocab) [inline Formula/cava.rb:21]: Looks like this `inreplace` broke the build on Intel too.

Comment (chincheta0815) [reply Formula/cava.rb]: Well, I did some digging:
It builds without issues locally when I set automake and libtool as "depends_on" without ":build"...
But I do not know why that is this way.

Shouldn't it work when just building?

File: Formula/cava.rb [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  depends_on "iniparser"
+  depends_on "portaudio"
+
+  uses_from_macos "ncurses"
+
+  def install
+    # force autogen.sh to look for and use our glibtoolize
+    inreplace "autogen.sh", "libtoolize", "glibtoolize"
+    # correct Makefile.am with hardcoded libpath
+    inreplace "Makefile.am", "/usr/local", "#{HOMEBREW_PREFIX}"
```
Comment (carlocab) [inline Formula/cava.rb:21]: This won't help and will just cause `brew style` errors.
```suggestion
    inreplace "Makefile.am", "/usr/local", HOMEBREW_PREFIX
```
What you need is a build dependency on `autoconf` since `autogen.sh` is looking for `autom4te` after `Makefile.am` has been modified.

Comment (chincheta0815) [reply Formula/cava.rb]: I changed that.

The thing with the build dependency: I tried that. But it fails. With a "standard" dependency it nevertheless works...

File: Formula/cava.rb (hunk 6 above)
Comment (carlocab) [reply Formula/cava.rb:21]: Does it stop working locally when you make them `:build` dependencies?

Comment (carlocab) [reply Formula/cava.rb:21]: You can make them `:build` dependencies if you also add a build dependency on `autoconf`.

File: Formula/cava.rb (hunk 7 above)
Comment (carlocab) [reply Formula/cava.rb]: If you add `:build` dependency on `autoconf` it will work.

File: Formula/cava.rb (hunk 6 above)
Comment (chincheta0815) [reply Formula/cava.rb:22]: ahh okay. :build dependencies have to list piece by piece... I thought since automake has a dependency on autoconf, all dependencies are there...

File: Formula/cava.rb (hunk 7 above)
Comment (chincheta0815) [reply Formula/cava.rb]: ahh okay. :build dependencies have to list piece by piece... I thought since automake has a dependency on autoconf, all dependencies are there...

File: Formula/cava.rb (hunk 6 above)
Comment (carlocab) [reply Formula/cava.rb:22]: Can we report the hardcoded `/usr/local` upstream? The `libtoolize` `inreplace` is fine, but they must support configurable prefixes.

**Comment by carlocab** [comment]
This is looking good. Thanks for your work here, @chincheta0815. I think we only have the test to sort out.

**Comment by chincheta0815** [comment]
> This is looking good. Thanks for your work here, @chincheta0815. I think we only have the test to sort out.

What do you mean by sort out? Delete it?

As written above this is the best test that I can get from the program.
The thing is that cava is a binary that gets an input audio stream and visualizes that in the terminal or by raw data.
For checking whether it works, there is a input stream needed and if this would be there, the ever-changing output has to be checked for.

Using the "version" or the "help" of the program as a command line args it as far as I see the best reproducable output we can get for a test.

File: Formula/cava.rb (hunk 6 above)
Comment (chincheta0815) [reply Formula/cava.rb:22]: I ask them.
The /usr/local is not referring to the prefix. This is set correctly.
It refers to the LDFLAGS.

File: Formula/cava.rb (hunk 4 above)
Comment (carlocab) [reply Formula/cava.rb:40]: > What do you mean by sort out?

See above. We can always dump all the output to `/dev/null` given that we're just checking that it doesn't error.

File: Formula/cava.rb (hunk 6 above)
Comment (carlocab) [reply Formula/cava.rb:22]: Yes, it doesn't refer to the prefix, but this also causes build failures on systems that don't have `/usr/local`.

**Comment by chincheta0815** [comment]
> > What do you mean by sort out?
> 
> You need to delete the comments, at the very least. See also the discussion at [#89820 (comment)](https://github.com/Homebrew/homebrew-core/pull/89820#discussion_r754250220) for other things we can do.

Okay. Comments are gone now.
Piping to /dev/null would mean that there is an input. As this would be audio I do not know if the build bot can play that...
In addition, cava can only be stopped by STRG^C.

So far I do not see a good stable and reproducable/predictive option for a test, although I'd like to have one indeed...

**Comment by chincheta0815** [comment]
It seems to build now correctly with the updated mainstream.
Nevertheless, I have to take care whether HEAD is installed or a versioned tag.
The former does not need the inreplace any more. The latter does until there is a new version being tagged...

Is my way of handling that okay?
If yes there remains the "test issue" and than I might rebase the commit again...

**Comment by SMillerDev** [comment]
> Piping to /dev/null would mean that there is an input. As this would be audio I do not know if the build bot can play that...

I guess there's only one way to find out.

> In addition, cava can only be stopped by STRG^C.

Does it not respond to SIGTERM?

**Comment by chincheta0815** [comment]
> > Piping to /dev/null would mean that there is an input. As this would be audio I do not know if the build bot can play that...
> 
> I guess there's only one way to find out.
Actually, I do not know this way.

> 
> > In addition, cava can only be stopped by STRG^C.
> 
> Does it not respond to SIGTERM?
For the src code it reacts on SIGUSR1, SIGUSR2, SIGINT.

In terms of proper testing one should be testing against something you would be waiting for.
Using audio, one would wait for a pulsing bar description drawn by chars in the terminal (e.g. 
(https://github.com/karlstav/cava/blob/master/example_files/cava.gif).
Another thing would be to start the portaudio defaults audio stream. That has to be started before cava is started...

Summed up: To my opinion "cava -v" is a good an feasible test. Other things seem to be a "piece of software" on their own...

**Comment by chincheta0815** [comment]
currently, there is still an issue with ncurses being not built correctly...

**Comment by carlocab** [comment]
> In terms of proper testing one should be testing against something you would be waiting for.

Why can't we try to run `cava` with more than just `-v` to check that it doesn't crash/segfault or otherwise produce an error? It would be nice to do all these other tests you suggest, but I agree with you that that would be quite complicated. Checking to see it doesn't crash is still a test that is feasible and is much better than just doing `-v`.

> currently, there is still an issue with ncurses being not built correctly...

What's wrong with `ncurses`?

**Comment by chincheta0815** [comment]
> Why can't we try to run `cava` with more than just `-v` to check that it doesn't crash/segfault or otherwise produce an error? It would be nice to do all these other tests you suggest, but I agree with you that that would be quite complicated. Checking to see it doesn't crash is still a test that is feasible and is much better than just doing `-v`.
> 
cava -p would expect a config file. when there is none is show the help which can be shown by "cava -h". Only difference to "cava -h" would be that cava tells something about the usage: "cava: option requires an argument -- p". 
would that be okay in addition to "cava -v"?
 
> What's wrong with `ncurses`?
Sorry, I was misleading.
cava is built without ncurses support. this is something I do not understand. I thought that "uses_from_macos ncurses", makes it compile against ncurses. trying to fix that I am now at the point that I do not find any ncurses lib in /usr/lib...
Do you have a hint?

**Comment by chincheta0815** [comment]
@carlocab Okay, concerning ncurses I got it fixed. 
cava expect ncursesw and macos seems to have just ncures.
Unfortunately, by using an inreplace that tweak the configure.ac before running autogen.sh.
That is not quite elegant.
Is that okay?

**Comment by SMillerDev** [comment]
Are there no functional differences between `ncursesw` and `ncurses`?

**Comment by chincheta0815** [comment]
> Are there no functional differences between `ncursesw` and `ncurses`?

As even the devs use ncurses instead of ncursesw for building by setting a link there should be no difference that effects a proper cava build. So far it works fine on my box. 

Here the link to cava's build instructions:
https://github.com/karlstav/cava#from-source

**Comment by carlocab** [comment]
> I do not find any ncurses lib in /usr/lib...

System libraries no longer live in the file system. They're in the dyld cache. The linker should just be able to find them with the `-l` flag. The issue usually is when the build system insists on checking the file system for them, which I imagine might be what's happening here.

> Are there no functional differences between `ncursesw` and `ncurses`?

There are a lot. They're not even binary compatible. But using `ncursesw` when you actually need `ncurses` kind of works.

> As even the devs use ncurses instead of ncursesw for building by setting a link there should be no difference that effects a proper cava build.

That's because macOS `ncurses` is actually `ncursesw`.

**Comment by chincheta0815** [comment]
Okay it tried to do a test with the option "-p". This would result in an error message stating -p requires a filename. 
This is an expected behaviour and can be tested. The only thing: the test fails as assert_match get an "error return code" from cava...

**Comment by chincheta0815** [comment]
> > I do not find any ncurses lib in /usr/lib...
> 
> System libraries no longer live in the file system. They're in the dyld cache. The linker should just be able to find them with the `-l` flag. The issue usually is when the build system insists on checking the file system for them, which I imagine might be what's happening here.
> 
> > Are there no functional differences between `ncursesw` and `ncurses`?
> 
> There are a lot. They're not even binary compatible. But using `ncursesw` when you actually need `ncurses` kind of works.
> 
> > As even the devs use ncurses instead of ncursesw for building by setting a link there should be no difference that effects a proper cava build.
> 
> That's because macOS `ncurses` is actually `ncursesw`.

Okay. The here everything is okay. At least for me ;o)

**Comment by carlocab** [comment]
We should probably just do `depends_on "ncurses"`. The one that comes with macOS is over a decade old and has poor support for modern terminals.

As for the test: I still don't understand why we can't do a proper test that does more than just `-p` or `-v` to check it doesn't crash. You can even [run `cava` over `ssh`](https://github.com/karlstav/cava#running-via-ssh), so it seems like running on headless CI shouldn't be a problem.

**Comment by chincheta0815** [comment]
> We should probably just do `depends_on "ncurses"`. The one that comes with macOS is over a decade old and has poor support for modern terminals.
> 
Done.

> As for the test: I still don't understand why we can't do a proper test that does more than just `-p` or `-v` to check it doesn't crash. You can even [run `cava` over `ssh`](https://github.com/karlstav/cava#running-via-ssh), so it seems like running on headless CI shouldn't be a problem.

cava has a "terminal gui (ncurses, curses, ...)". it is no command line tool, it simplay shows equalizer bars in the terminal. while doing that you can "steer" it with some keyboards keys.
Running it over ssh means you start it on the remote computer and redirect the output on the remote computer to its console to show in the monitor screen.

The only test that worked well so far was for "-v". With "-p" one can check if cava would "stop with an error" as expected.

I do not have any other ideas how to test that properly... 
Is there a way that you probably install cava and give it a try as it is hard for me to explain more to that? The doc of cava is explaining very good what cava does...

**Comment by chincheta0815** [comment]
> > We should probably just do `depends_on "ncurses"`. The one that comes with macOS is over a decade old and has poor support for modern terminals.
> 
Using "depends_on ncurses" fails the audit, recommending using the one from macos.

**Comment by chincheta0815** [comment]
@carlocab This is the latest version of the formula. after discussing with the devs they came up with an idea concerning a test that is now implemented. the thing you mentioned concerning ncurses did not work as I get the message from the audit that I should use "uses_from_macos" rather than "depends_on". right now the testbot seems to hang, but some tests already passed...

File: Formula/cava.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      [output]
+      method = raw
+      data_format = ascii
+    EOS
+
+    require "pty"
+    PTY.spawn("#{bin}/cava -p #{cava_config} > #{cava_stdout}") do |_stdout, _stdin, pid|
+      sleep 2
+      Process.kill "KILL", pid
+    end
```
Comment (carlocab) [inline Formula/cava.rb:56]:
```suggestion
    pid = spawn(bin/"cava", "-p", cava_config, out: cava_stdout)
    sleep 2
    Process.kill "KILL", pid
```
Let's use `Kernel#spawn` which can handle redirection on its own.

File: Formula/cava.rb [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    log_content = cava_stdout.read
+    assert_match "0;0;\n", log_content
```
Comment (carlocab) [inline Formula/cava.rb:58]:
```suggestion
    assert_match "0;0;\n", cava_stdout.read
```

File: Formula/cava.rb [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+  head "https://github.com/karlstav/cava.git", branch: "master"
+
+  depends_on "autoconf" => :build
+  depends_on "automake" => :build
+  depends_on "libtool"  => :build
+  depends_on "fftw"
+  depends_on "iniparser"
+  depends_on "portaudio"
+
+  uses_from_macos "ncurses"
```
Comment (carlocab) [inline Formula/cava.rb:16]: Didn't you say this doesn't link with macOS `ncurses`? We can use Homebrew `ncurses` and ignore the audit failure.

File: Formula/cava.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+  def install
+    # change ncursesw to ncurses
+    inreplace "configure.ac", "ncursesw", "ncurses"
```
Comment (carlocab) [inline Formula/cava.rb:20]: If we use Homebrew `ncurses` we won't need this.

**Review by carlocab — COMMENTED** [review]
Thanks; the test looks great! Just a few more comments, but this should be ready to merge soon.

As for CI: it didn't hang. There's just a very long queue for the runners (we only have a limited number of them).

File: Formula/cava.rb (hunk 10 above)
Comment (chincheta0815) [reply Formula/cava.rb:16]: It links with macos ncurses as long as the inreplace takes place. For avoiding the inreplace in the future the devs already pushed to master. But version 0.7.4 still uses this inreplace. I made this inreplace conditional.

File: Formula/cava.rb (hunk 11 above)
Comment (chincheta0815) [reply Formula/cava.rb:20]: I have to problem using homebrew ncurses. But how do I use this?
When I add a "depends_on ncurses" the testbot tells me to use "uses_from_macos"...

File: Formula/cava.rb (hunk 10 above)
Comment (carlocab) [reply Formula/cava.rb:16]: Ok, I see. If it works fine with macOS ncurses then we can leave it.

File: Formula/cava.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+      [input]
+      method = fifo
+      source = /dev/zero
+
+      [output]
+      method = raw
+      data_format = ascii
+    EOS
+
+    pid = spawn("#{bin}/cava -p #{cava_config}", [:out, :err] => cava_stdout.to_s)
```
Comment (carlocab) [inline Formula/cava.rb:52]:
```suggestion
    pid = spawn(bin/"cava", "-p", "cava_config", [:out, :err] => cava_stdout.to_s)
```
Please don't pass this as one string. See my previous suggestion.

Comment (chincheta0815) [reply Formula/cava.rb:52]: Sorry.

**Review by carlocab — APPROVED** [review]
Thanks for sticking with this, @chincheta0815! Nice work.
